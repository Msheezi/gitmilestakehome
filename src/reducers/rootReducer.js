import {combineReducers} from 'redux'
import undoableRewards from './rewardsReducer'


const rootReducer = combineReducers({
  rewards: undoableRewards,
});

export default rootReducer