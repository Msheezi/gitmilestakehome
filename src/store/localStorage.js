export const loadState = () => {
    try {
        const serializedState = localStorage.getItem('savedState')
        if (serializedState === null){
            return undefined 
        }
        return {rewards:JSON.parse(serializedState)}
    } catch (err){
        return undefined
    }
}

export const saveState = state => {
    try {
        const serializedState = JSON.stringify(state)
        localStorage.setItem("savedState", serializedState)
    } catch {
        //ignore write errors
    }
}