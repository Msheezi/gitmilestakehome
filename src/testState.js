export const testState = {
         rewards: {
           past: [
             [],
             [
               {
                 name: "r1",
                 location: "r1c1",
                 bgcolor: "green",
                 key: "1592131225077",
               },
             ],
           ],
           present: [
             {
               name: "r1",
               location: "r1c1",
               bgcolor: "green",
               key: "1592131225077",
             },
             {
               name: "r2",
               location: "r2c2",
               bgcolor: "yellow",
               key: "1592131226860",
             },
           ],
           future: [],
           group: null,
           _latestUnfiltered: [
             {
               name: "r1",
               location: "r1c1",
               bgcolor: "green",
               key: "1592131225077",
             },
             {
               name: "r2",
               location: "r2c2",
               bgcolor: "yellow",
               key: "1592131226860",
             },
           ],
           index: 2,
           limit: 3,
         },
       };
