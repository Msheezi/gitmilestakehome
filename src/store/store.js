import { createStore, applyMiddleware } from "redux";

import logger from "redux-logger";

import RewardsReducer from "../reducers/rewardsReducer";
import { loadState } from './localStorage'



const configureStore = () => {
  const persistedState = loadState()
  return createStore(RewardsReducer, persistedState, applyMiddleware(logger));

}

 
export default configureStore;
