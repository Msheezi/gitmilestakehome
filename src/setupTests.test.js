// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent } from "@testing-library/react";
import {Provider} from 'react-redux'
import React from "react";
import configureStore2 from 'redux-mock-store'
import renderer from 'react-test-renderer'

import ReactDOM from "react-dom";
import DragAndDrop from "./components/draganddrop";
import UndoRedo from './components/UndoRedo'
import configureStore from "./store/store";
import rootReducer from './reducers/rootReducer';
import { saveState } from './store/localStorage';
import { createStore } from 'redux';
import {testState} from './testState'

const mockStore = configureStore2([])

const renderWithRedux = (
  component, 
  { initialState,
     store = createStore(rootReducer, initialState)} ={}
) => {
return {
  ...render(<Provider store={store}>{component}</Provider>),
store,
}
}

test("It renders without Crashing", () => {
  const div = document.createElement("div");
  const store = configureStore();
  store.subscribe(()=> {
    saveState.getState()
  })
  ReactDOM.render(<DragAndDrop store={store} />, div);
  //   const { getByText } = render(<DragandDrop />);

  //   const linkElement = getByText(/learn react/i);
  //   expect(linkElement).toBeInTheDocument();
});

test("Reward r1 renders", () => {
  const { getByTestId } = renderWithRedux(<DragAndDrop />);
  expect(getByTestId("rewardsr1"));
});

test("Reward r2 renders", () => {
  const { getByTestId } = renderWithRedux(<DragAndDrop />);
  expect(getByTestId("rewardsr1"));
});

test("Reward r3 renders", () => {
  const { getByTestId } = renderWithRedux(<DragAndDrop />);
  expect(getByTestId("rewardsr1"));
});

test("Reward r4 renders", () => {
  const { getByTestId } = renderWithRedux(<DragAndDrop />);
  expect(getByTestId("rewardsr1"));
});

test("Reward r5 renders", () => {
  const { getByTestId } = renderWithRedux(<DragAndDrop />);
  expect(getByTestId("rewardsr1"));
});



describe("Stored Rewards Render Correctly", () => {
  let store 
  let component

  beforeEach(()=> {
    store = mockStore({
      rewards: {
        present: [
          {
            name: "r5",
            location: "r5c1",
            bgcolor: "orange",
            key: "1592096048040",
          },
        ],
        future: [],
        group: null,
        _latestUnfiltered: [
          {
            name: "r5",
            location: "r5c1",
            bgcolor: "orange",
            key: "1592096048040",
          },
        ],
        index: 44,
        limit: 45,
      },
    });
    component = renderer.create(
      <Provider store={store}>
        <DragAndDrop/>
      </Provider>
    )
  })
    
  it('should render with given state', () =>{
    expect(component.toJSON()).toMatchSnapshot()
  })

  // it('should place rewards when dropped in swimlane', ()=>{
  //   renderer.act(()=> {
  //     component.root.findByType
  //   })
  // })
})

describe("Should allow User to Undo Actions", ()=> {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      rewards: {
        past: [
          [],
          [
            {
              name: "r1",
              location: "r1c1",
              bgcolor: "green",
              key: "1592131225077",
            },
          ],
        ],
        present: [
          {
            name: "r1",
            location: "r1c1",
            bgcolor: "green",
            key: "1592131225077",
          },
          {
            name: "r2",
            location: "r2c2",
            bgcolor: "yellow",
            key: "1592131226860",
          },
        ],
        future: [],
        group: null,
        _latestUnfiltered: [
          {
            name: "r1",
            location: "r1c1",
            bgcolor: "green",
            key: "1592131225077",
          },
          {
            name: "r2",
            location: "r2c2",
            bgcolor: "yellow",
            key: "1592131226860",
          },
        ],
        index: 2,
        limit: 3,
      },
    });
    component = renderer.create(
      <Provider store={store}>
       <DragAndDrop />
        <UndoRedo/>
      </Provider>
    );
  })
  
  it('Should Allow Undo', ()=>{
    renderer.act(()=>{
      component.root.findByProps({type:"undo"}).props.onCLlick()
    })
    expect(store.dispatch).toHaveBeenCalledTimes(1)
  })
})

