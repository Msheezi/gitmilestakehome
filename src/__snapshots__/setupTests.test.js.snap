// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stored Rewards Render Correctly should render with given state 1`] = `
<div
  className="sc-AxjAm bhEZON"
>
  <div
    className="sc-Axmtr zEjJq"
  />
  <h2
    className="sc-AxiKw dMAqCs"
  >
    Rewards
  </h2>
  <div
    className="sc-AxirZ kLgyuO"
    color="#f5e6e8"
    data-testid="rewardsr1"
    draggable={true}
    onDragStart={[Function]}
  >
    r1
  </div>
  <div
    className="sc-AxirZ fUVzAA"
    color="#a3f7b5"
    data-testid="rewardsr2"
    draggable={true}
    onDragStart={[Function]}
  >
    r2
  </div>
  <div
    className="sc-AxirZ bOZNwc"
    color="#40c9a2"
    data-testid="rewardsr3"
    draggable={true}
    onDragStart={[Function]}
  >
    r3
  </div>
  <div
    className="sc-AxirZ dlAjTo"
    color="#2f9c95"
    data-testid="rewardsr4"
    draggable={true}
    onDragStart={[Function]}
  >
    r4
  </div>
  <div
    className="sc-AxirZ libgMC"
    color="#bee3ea"
    data-testid="rewardsr5"
    draggable={true}
    onDragStart={[Function]}
  >
    r5
  </div>
  <div
    className="sc-AxhCb jFVYKg"
  >
    <h2
      className="sc-AxiKw jaxNuQ"
    >
      Categories
    </h2>
  </div>
  <div
    className="sc-AxhUy cvQYZU rewards1"
    data-testid="c1Dropzone"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span
      className="category"
    >
      c1
    </span>
  </div>
  <div
    className="sc-AxhUy hxGMLf rewards1"
    data-testid="c2Dropzone"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span
      className="category"
    >
      c2
    </span>
  </div>
  <div
    className="sc-AxhUy iNmKoq rewards1"
    data-testid="c3Dropzone"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span
      className="category"
    >
      c3
    </span>
  </div>
  <div
    className="sc-AxhUy dqnlUN rewards1"
    data-testid="c4Dropzone"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span
      className="category"
    >
      c4
    </span>
  </div>
  <div
    className="sc-AxhUy dpMtoM rewards1"
    data-testid="c5Dropzone"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <span
      className="category"
    >
      c5
    </span>
  </div>
  <div
    className="sc-AxgMl iTkSug"
    color="orange"
    data-testid="reward r5c1"
    draggable={true}
    onDragStart={[Function]}
  >
    r5
    <div
      className="sc-AxheI gEUzsA"
      onClick={[Function]}
    >
      X
    </div>
  </div>
  <div
    className="sc-AxmLO eyIhGq"
  />
</div>
`;
